lib/cnet/punt/kernel_recv.c

```diff
@@ -200,40 +231,59 @@ kernel_recv_node_init(const struct cne_graph *graph __cne_unused, struct cne_nod
     pktmbuf_info_t *pi;
     mmap_t *mm;
 
-    if (!ctx)
-        CNE_ERR_RET("Node context pointer is NULL\n");
+    ctx->recv_info = calloc(1, sizeof(kernel_recv_info_t));
+    if (!ctx->recv_info)
+        CNE_ERR_RET("Kernel recv_info is NULL\n");
 
-    ctx->tinfo = tun_alloc(IFF_TUN | IFF_NO_PI, "krecv%d");
-    if (!ctx->tinfo)
-        CNE_ERR_RET("Unable to open TUN/TAP socket\n");
+    ctx->sock = socket(AF_INET, SOCK_RAW, IPPROTO_RAW);
+    if (ctx->sock < 0)
+        CNE_ERR_RET("Unable to open RAW socket\n");
 
     /* Allocate a memory region and pktmbuf info structure to be used in receiving
      * packets from the kernel */
     mm = mmap_alloc(KERN_RECV_MBUF_COUNT, DEFAULT_MBUF_SIZE, MMAP_HUGEPAGE_DEFAULT);
     if (!mm) {
-        tun_free(ctx->tinfo);
+        close(ctx->sock);
+        ctx->sock = -1;
         CNE_ERR_RET("mmap_alloc() failed\n");
     }
 
     pi = pktmbuf_pool_create(mmap_addr(mm), KERN_RECV_MBUF_COUNT, DEFAULT_MBUF_SIZE, 0, NULL);
     if (!pi) {
         mmap_free(mm);
-        tun_free(ctx->tinfo);
+        close(ctx->sock);
+        ctx->sock = -1;
         CNE_ERR_RET("pktmbuf_pool_create() failed\n");
     }
-    pktmbuf_info_name_set(pi, tun_get_name(ctx->tinfo));
-    ctx->pi = pi;
+    pktmbuf_info_name_set(pi, graph->name);
+    ctx->recv_info->pi = pi;
+    ctx->recv_info->mm = mm;
 
-    tun_dump(NULL, ctx->tinfo);
     return 0;
 }
 
+static void
+kernel_recv_node_fini(const struct cne_graph *graph __cne_unused, struct cne_node *node)
+{
+    kernel_recv_node_ctx_t *ctx = (kernel_recv_node_ctx_t *)node->ctx;
+
+    close(ctx->sock);
```

Comment:
I looked through the graph code, and the de-constructor is only called if the (all the nodes') constructor succeeded, when the graph is destroyed. This means that ctx->sock will reference a valid socket.

But there is a different problem, where if we have some nodes that fail to initialize later (after this one) we will never actually call any of the de constructors, so we'll never close this or free memory, etc. We should fix that at some point (different PR).

Reply:
Yes, I agree. I should have tested for ctx->sock >= 0 ??

Reply:
If we thought `fini` were called even when `init` fails, yes, but that isn't how the graph code works. It only calls `fini` functions when all the `init` functions in the graph succeed. Therefore we know `ctx->sock` will be valid, otherwise it would have failed in `init` and the `fini` functions would never be called (which is a different memory/resource leak problem)